File: src/Sequence/content/SqOverride.java
package Sequence.content;

import Sequence.core.SeqElem;
import Sequence.core.SqBundle;
import Sequence.core.SqEventType;
import Sequence.core.SqTmp;
import Sequence.ui.SqContentInfoDialog;
import Sequence.world.meta.SqStat;
import Sequence.world.util.Change;
import Sequence.world.util.Util;
import arc.Events;
import arc.struct.ObjectMap;
import mindustry.content.Blocks;
import mindustry.ctype.UnlockableContent;
import mindustry.entities.bullet.BulletType;
import mindustry.game.EventType;
import mindustry.world.blocks.defense.turrets.ItemTurret;
import mindustry.world.blocks.defense.turrets.Turret;

import static mindustry.Vars.content;
import static mindustry.Vars.ui;

public class SqOverride {
    public static void setup() {
        Events.on(EventType.ClientLoadEvent.class, ignore -> {
            ui.content = new SqContentInfoDialog();

            for (UnlockableContent content : content.blocks()) {
                if (content instanceof Turret turret) {
                    if (turret instanceof ItemTurret itemTurret) {
                        itemTurret.buildType = () -> itemTurret.new ItemTurretBuild() {
                            public boolean consumeLiquidAndMultiShoot(SqLiquids.VectorizedFluid liq, float amount, BulletType type) {
                                if (liquids.get(liq) > amount) {
                                    liquids.remove(liq, amount);
                                    float multi = liq.damageMulti;
                                    float knockbackMulti = liq.knockbackMulti;
                                    Util.checkKey(
                                            Util.checkKey(
                                                    Util.checkKey(
                                                            SqTmp.damageMultiMap,
                                                            type,
                                                            ObjectMap::new
                                                    ),
                                                    multi,
                                                    ObjectMap::new
                                            ),
                                            knockbackMulti,
                                            () -> Change.bulletType(type.copy(), bt -> {
                                                bt.damage *= multi;
                                                bt.splashDamage *= multi;
                                                bt.lightningDamage *= multi;
                                                bt.knockback *= knockbackMulti;
                                            })
                                    );
                                    super.shoot(SqTmp.damageMultiMap.get(type).get(multi).get(knockbackMulti));
                                    return true;
                                }
                                return false;
                            }

                            @Override
                            protected void shoot(BulletType type) {
                                for (SqLiquids.VectorizedFluid fluid : SqLiquids.VectorizedFluid.all)
                                    if (consumeLiquidAndMultiShoot(fluid, fluid.consumeAmount, type))
                                        return;
                                super.shoot(type);
                            }
                        };
                    }
                }
            }

            Events.on(SqEventType.ContentStatInitEvent.class, e -> {
                if (e.content instanceof SeqElem elem) {
                    e.content.stats.useCategories = true;
                    e.content.stats.add(SqStat.sequenceOrder, elem.order() == -1 ? SqBundle.modCat("seq-null") : String.valueOf(elem.order()));
                    if (elem.order() != -1 && elem.statValue() != null) {
                        e.content.stats.add(SqStat.sequenceEffect, elem.statValue());
                    }
                }
            });

            Blocks.itemSource.health = Integer.MAX_VALUE;
            Blocks.liquidSource.health = Integer.MAX_VALUE;
            Blocks.powerSource.health = Integer.MAX_VALUE;
            Blocks.payloadSource.health = Integer.MAX_VALUE;
            Blocks.heatSource.health = Integer.MAX_VALUE;
            Blocks.itemVoid.health = Integer.MAX_VALUE;
            Blocks.liquidVoid.health = Integer.MAX_VALUE;
            Blocks.powerVoid.health = Integer.MAX_VALUE;
            Blocks.payloadVoid.health = Integer.MAX_VALUE;
        });
    }
}

File: src/sequence/content/SqOverride.kt
package sequence.content

import arc.Events
import mindustry.content.Blocks
import mindustry.ctype.UnlockableContent
import mindustry.game.EventType.ClientLoadEvent
import mindustry.world.blocks.defense.turrets.ItemTurret
import sequence.graphic.SqColor

object SqOverride {
    fun setup() {
        Events.on(ClientLoadEvent::class.java) {
            SqContentMap.init()
            Blocks.itemSource.health = Int.MAX_VALUE
            Blocks.liquidSource.health = Int.MAX_VALUE
            Blocks.powerSource.health = Int.MAX_VALUE
            Blocks.payloadSource.health = Int.MAX_VALUE
            Blocks.heatSource.health = Int.MAX_VALUE
            Blocks.itemVoid.health = Int.MAX_VALUE
            Blocks.liquidVoid.health = Int.MAX_VALUE
            Blocks.powerVoid.health = Int.MAX_VALUE
            Blocks.payloadVoid.health = Int.MAX_VALUE

            Blocks.foreshadow.override<ItemTurret> {
                ammoTypes.put(SqItems.grainBoundaryAlloy, SqBulletTypes.foreshadowGBA)
            }
        }
    }

    fun addOverrideTag(content: UnlockableContent) {
        content.description = content.description ?: ""
        content.description += "\n[#${SqColor.LiuDai}]Override by [#${SqColor.NedKelly}]Sequence[]"
    }

    inline fun <reified T : UnlockableContent> UnlockableContent.override(builder: T.() -> Unit) {
        addOverrideTag(this)
        builder(this as T)
    }
}
